### PR TITLE
mds: more unique_pointer changes

### PR DESCRIPTION
**Pull request by batrick**

### CODE REVIEW

**Comment by ukernel** [comment]
lgtm

File: src/mds/mdstypes.cc [hunk 1]
```diff
@@ -234,9 +234,9 @@ void inline_data_t::decode(bufferlist::iterator &p)
   uint32_t inline_len;
   ::decode(inline_len, p);
   if (inline_len > 0)
-    ::decode_nohead(inline_len, get_data(), p);
+    ::decode_nohead(inline_len, *blp, p);
```
Comment (jcsp) [inline src/mds/mdstypes.cc]: You have to call get_data, or you get a null pointer here and lots of crashing

**Comment by batrick** [comment]
@jcsp done! Thanks.

File: src/mds/mdstypes.h [hunk 2]
```diff
@@ -415,41 +415,37 @@ inline bool operator==(const client_writeable_range_t& l,
...
-  inline_data_t() : blp(0), version(1) {}
-  inline_data_t(const inline_data_t& o) : blp(0), version(o.version) {
+  inline_data_t() : version(1) {}
+  inline_data_t(const inline_data_t& o) : version(o.version) {
     if (o.blp)
-      get_data() = *o.blp;
-  }
-  ~inline_data_t() {
-    free_data();
+      *blp.get() = *o.blp.get();
```
Comment (jcsp) [inline src/mds/mdstypes.h]: Oops, didn't spot this until I ran the tests again, there's another crasher here, blp may be null at this point (shoudl use get_data() like before).  Same  in the copy constructor.

Comment (batrick) [inline src/mds/mdstypes.h]: Ugh, sorry. I just pushed the change.

**Comment by jcsp** [comment]
```
In file included from /srv/autobuild-ceph/gitbuilder.git/build/rpmbuild/BUILD/ceph-11.0.2/src/cls/cephfs/cls_cephfs_client.h:3:0,
from /srv/autobuild-ceph/gitbuilder.git/build/rpmbuild/BUILD/ceph-11.0.2/src/cls/cephfs/cls_cephfs_client.cc:16:
/srv/autobuild-ceph/gitbuilder.git/build/rpmbuild/BUILD/ceph-11.0.2/src/mds/mdstypes.h: In copy constructor 'inline_data_t::inline_data_t(const inline_data_t&)':
error: /srv/autobuild-ceph/gitbuilder.git/build/rpmbuild/BUILD/ceph-11.0.2/src/mds/mdstypes.h:435:31: passing 'const inline_data_t' as 'this' argument of 'ceph::bufferlist& inline_data_t::get_data()' discards qualifiers [-fpermissive]
get_data() = o.get_data();
^
/srv/autobuild-ceph/gitbuilder.git/build/rpmbuild/BUILD/ceph-11.0.2/src/mds/mdstypes.h: In member function 'inline_data_t& inline_data_t::operator=(const inline_data_t&)':
error: /srv/autobuild-ceph/gitbuilder.git/build/rpmbuild/BUILD/ceph-11.0.2/src/mds/mdstypes.h:440:31: passing 'const inline_data_t' as 'this' argument of 'ceph::bufferlist& inline_data_t::get_data()' discards qualifiers [-fpermissive]
get_data() = o.get_data();
```

**Comment by dachary** [comment]
jenkins test this please (no log)

**Comment by dachary** [comment]
<pre>
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/MDSMap.h:34:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/FSMap.h:24,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/MDSMonitor.h:26,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/Monitor.cc:72:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/mdstypes.h: In copy constructor ‘inline_data_t::inline_data_t(const inline_data_t&)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/mdstypes.h:435:31: error: passing ‘const inline_data_t’ as ‘this’ argument of ‘ceph::bufferlist& inline_data_t::get_data()’ discards qualifiers [-fpermissive]
       get_data() = o.get_data();
                               ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/mdstypes.h: In member function ‘inline_data_t& inline_data_t::operator=(const inline_data_t&)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mds/mdstypes.h:440:31: error: passing ‘const inline_data_t’ as ‘this’ argument of ‘ceph::bufferlist& inline_data_t::get_data()’ discards qualifiers [-fpermissive]
       get_data() = o.get_data();
</pre>

**Comment by batrick** [comment]
@jcsp I think this one is ready. Sorry about the silly bugs.

**Comment by jcsp** [comment]
@batrick this needs a quick rebase (i've hand merged for my test branch)

**Comment by batrick** [comment]
@jcsp done!

File: src/mds/MDCache.cc [hunk 3]
```diff
@@ -3322,8 +3313,8 @@ void MDCache::maybe_resolve_finish()
     assert(mds->is_resolve());
     trim_unlinked_inodes();
     recalc_auth_bits(false);
-    resolve_done->complete(0);
-    resolve_done = NULL;
+    assert(resolve_done);
```
Comment (jcsp) [inline src/mds/MDCache.cc]: This assertion is now redundant because we're inside an if(resolve_done)

Comment (batrick) [inline src/mds/MDCache.cc]: Fixed

**Pull request merged**